### PR TITLE
PERF: Remove unnecessary copies in sorting functions (#33917)

### LINKED ISSUE #33917: Performance regression in timeseries.SortIndex.time_sort_index
```python
import pandas as pd
import numpy as np

N = 10 ** 5
idx = pd.date_range(start="1/1/2000", periods=N, freq="s")
s = pd.Series(np.random.randn(N), index=idx)
%timeit s.sort_index()
```

```
# 1.0.2
108 µs ± 8.27 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
# master
225 µs ± 8.36 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

According to https://pandas.pydata.org/speed/pandas/index.html#timeseries.SortIndex.time_sort_index?p-monotonic=True&commits=f683473a156f032a64a1d7edcebde21c42a8702d-085860a49f3a87aa4e24b3115b50b85c4b3c5676, the first slow commit is https://github.com/pandas-dev/pandas/pull/33755, which just bumps Cython in numpydev... So probably not actually that commit.

### PR DESCRIPTION
**Pull request by mproszewska**
- [x] closes #33917
- [x] tests passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

This PR fixes pereformance regression after commit `dec736f3f` by removing unnecessary copies if `key=None` in sorting functions.

```
setup = """
import pandas as pd
import numpy as np
N = 10 ** 5
idx = pd.date_range(start="1/1/2000", periods=N, freq="s")
s = pd.Series(np.random.randn(N), index=idx)
"""

import timeit
timeit.timeit("s.sort_index()",setup=setup, number=100000)

# master
# 41.245621485984884
# now
# 14.826273362035863
```

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
can you add an asv which hits this. I think this regression is in 1.1 so no whatsnew is needed.

**Review by jreback — CHANGES_REQUESTED** [review]
also in tests where this is hit, can you assert that values is not modified

**Comment by mproszewska** [comment]
> also in tests where this is hit, can you assert that values is not modified

This is used in function that sorts series by index, so I believe that values=index will be modified

File: asv_bench/benchmarks/algorithms.py [hunk 1]
```diff
@@ -174,4 +174,14 @@ def time_argsort(self, N):
...
+class SortIndexSeries:
+    def setup(self):
+        N = 10 ** 5
+        idx = pd.date_range(start="1/1/2000", periods=N, freq="s")
+        self.s = pd.Series(np.random.randn(N), index=idx)
+
+    def time_sort_index(self):
+        self.s.sort_index()
+
+
```
Comment (TomAugspurger) [inline asv_bench/benchmarks/algorithms.py:186]: I don't think we need this. SortIndex.time_sort_index already covers this code path.

**Comment by TomAugspurger** [comment]
Can you add a release note stating that we fixed the performance regression in sort_index?

**Comment by mproszewska** [comment]
> Can you add a release note stating that we fixed the performance regression in sort_index?

Should I add note even though regression was in 1.1?

File: asv_bench/benchmarks/algorithms.py (hunk 1 above)
Comment (mproszewska) [reply asv_bench/benchmarks/algorithms.py]: Right, I haven't checked `timeseries` before adding asv

**Comment by jreback** [comment]
> > Can you add a release note stating that we fixed the performance regression in sort_index?
> 
> Should I add note even though regression was in 1.1?

no this is fine (just add the asvs)

**Comment by mproszewska** [comment]
> > > Can you add a release note stating that we fixed the performance regression in sort_index?
> > 
> > 
> > Should I add note even though regression was in 1.1?
> 
> no this is fine (just add the asvs)

Isn't `timeseries.SortIndex.time_sort_index` enough?

**Comment by jreback** [comment]
what does it show pls post a run

**Comment by mproszewska** [comment]
```
     <perf>           <master>  
+        158±50μs         367±30μs     2.32  timeseries.SortIndex.time_sort_index(True)
+      3.84±0.4ms       5.94±0.5ms     1.55  timeseries.DatetimeAccessor.time_dt_accessor_normalize(None)
-      3.31±0.1ms       2.47±0.3ms     0.75  timeseries.ResampleDataFrame.time_method('max')
```

**Review by jreback — APPROVED** [review]
looks fine

can u merge master and ping on green

**Comment by mproszewska** [comment]
> looks fine
> 
> can u merge master and ping on green

ping

**Comment by jreback** [comment]
thanks @mproszewska

**Pull request merged**